### PR TITLE
Cleanup of string initialization code for limited API

### PR DESCRIPTION
**Pull request by da-woods**
The initialization now appears in a few large blocks rather than many

```
 # if CYTHON_LIMITED_API
   individual string line
 # endif
```

-----------------------

On the fairly simple `compute_prange` example (arbitrary example, chosen just because that's what I was looking at at the time) it saves about 900 lines of preprocessor statements (about 3% of the total file length).

See #2542

### CODE REVIEW

File: Cython/Compiler/Code.py [hunk 1]
```diff
@@ -1494,6 +1494,23 @@ def generate_string_constants(self):
             w = self.parts['pystring_table']
             w.putln("")
             w.putln("static __Pyx_StringTabEntry %s[] = {" % Naming.stringtab_cname)
+            if py_string.py3str_cstring:
+                w.putln("#if PY_MAJOR_VERSION >= 3")
+                w_py3str_writer = w.insertion_point()
+                w.putln("#else")
+            w.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
```
Comment (scoder) [inline Cython/Compiler/Code.py]: This seems wrong. Only Py3 has the limited API.

File: Cython/Compiler/Code.py [hunk 2]
```diff
@@ -1494,6 +1494,23 @@ def generate_string_constants(self):
             w = self.parts['pystring_table']
             w.putln("")
             w.putln("static __Pyx_StringTabEntry %s[] = {" % Naming.stringtab_cname)
+            if py_string.py3str_cstring:
```
Comment (scoder) [inline Cython/Compiler/Code.py]: `py_string` is a left-over loop variable here that may or may not have been assigned.

File: Cython/Compiler/Code.py (hunk 1 above)
Comment (da-woods) [reply Cython/Compiler/Code.py]: It is (but it's also wrong in master). For example the test "cython3_no_unicode_literals" crashes with -DCYTHON_LIMITED_API. I'll get this fixed though

**Comment by da-woods** [comment]
This now fixes a limited API bug. Previously the following file (a cut-down version of the "cython3_no_unicode_literals" test) failed to compile, and now it will run successfully.

```
# cython: language_level=3str, binding=True

def non_ascii_str():
    s = 'ø\x20\u0020'
    assert isinstance(s, str)
    return s
```
-----

The full "cython3_no_unicode_literals" test currently segfaults with the limited API (but is excluded by `--no-file` anyway), but I don't think that is related to this change since it happens on master too (with the offending function that causes the compile error removed)

File: Cython/Compiler/Code.py [hunk 3]
```diff
@@ -1512,32 +1523,27 @@ def generate_string_constants(self):
...
                         ))
-                    w.putln("#else")
-
-                w.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
-                w.putln("{0, %s, sizeof(%s), %s, %d, %d, %d}," % (
-                    c_cname,
-                    c_cname,
-                    encoding,
+                    w_not_limited_writer.putln("#else")
+                w_limited_writer.putln("{0, %s, sizeof(%s), %s, %d, %d, %d}," % (
```
Comment (scoder) [inline Cython/Compiler/Code.py]: Could you move this up or down, so that the whole non-limited API code is generated together? It's a bit confusing to see the preprocessor guards surround unrelated code in the code generator flow.

**Comment by scoder** [comment]
Looks good, thanks!
